### DISCUSSION TITLE
Where are the key source codes for the S3TransferManager.downloadFile() and S3TransferManager.copy() methods

### DISCUSSION THREAD

**Discussion by sokui** [aws/aws-sdk-java-v2#3489] [Q&A]
Hello community,

I am new to this lib. I tested the S3TransferManager.downloadFile() and S3TransferManager.copy() APIs, and found they are much faster than the old TransferManager. I am interested in looking at the source code of how these two methods are implemented. However, the source code is written in a so abstract way that I am totally lost. Can anyone help to point me out where is the real mechanism part in the code for these two methods? Really appreciate it!

**Comment by sokui** [comment]
I believe the S3TransferManager.downloadFile() must use byte range get to concurrently download different parts of the object into the local file to accelerate download speed. Please let me know where the code is. Thank you!

**Reply by debora-ito** [reply]
[`DefaultS3TransferManager`](https://github.com/aws/aws-sdk-java-v2/blob/master/services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/DefaultS3TransferManager.java) is the implementation of the `S3TransferManager` interface, from there you can follow the deeper implementations using the S3CrtAsyncClient, depending of what you are looking for.

**Reply by sokui** [reply]
@debora-ito , Thanks for your answer. I am looking into the detail implementation for downloading object from S3. Just try to understand why this lib is faster than the previous version. I expect to see there are concurrent byte range gets going on to download an object to a local file. I tracked down to DefaultS3AsyncClient. But I got lost. Could you please point me out where this code (probably some concurrent inputStreams) is?
